kubespawner/spawner.py

```diff
@@ -1774,72 +1876,31 @@ def _start(self):
-                    t, v, tb = sys.exc_info()
-
-                    try:
-                        yield self.asynchronize(
-                            self.api.read_namespaced_persistent_volume_claim,
-                            name=self.pvc_name,
-                            namespace=self.namespace)
-
-                    except ApiException as e:
-                        raise v.with_traceback(tb)
-
-                    self.log.info("PVC " + self.pvc_name + " already exists, possibly have reached quota though.")
-
-                else:
-                    raise
-
+            # If there's a timeout, just let it propagate
+            yield exponential_backoff(
+                partial(self._make_create_pvc_request, pvc, self.k8s_api_request_timeout),
+                f'Could not create pod {self.pvc_name}',
+                # Each req should be given k8s_api_request_timeout seconds.
+                # FIXME: We should instead add a timeout_times property to exponential_backoff instead
+                timeout=self.k8s_api_request_timeout * self.k8s_api_request_timeout_retries
+            )
         # If we run into a 409 Conflict error, it means a pod with the
         # same name already exists. We stop it, wait for it to stop, and
         # try again. We try 4 times, and if it still fails we give up.
-        # FIXME: Have better / cleaner retry logic!
-        retry_times = 4
         pod = yield self.get_pod_manifest()
         if self.modify_pod_hook:
             pod = yield gen.maybe_future(self.modify_pod_hook(self, pod))
-        for i in range(retry_times):
-            try:
-                yield self.asynchronize(
-                    self.api.create_namespaced_pod,
-                    self.namespace,
-                    pod,
-                )
-                break
-            except ApiException as e:
-                if e.status != 409:
-                    # We only want to handle 409 conflict errors
-                    self.log.exception("Failed for %s", pod.to_str())
-                    raise
-                self.log.info('Found existing pod %s, attempting to kill', self.pod_name)
-                # TODO: this should show up in events
-                yield self.stop(True)
-
-                self.log.info('Killed pod %s, will try starting singleuser pod again', self.pod_name)
-        else:
-            raise Exception(
-                'Can not create user pod %s already exists & could not be deleted' % self.pod_name)
+
+        # If there's a timeout, just let it propagate
+        yield exponential_backoff(
+            partial(self._make_create_pod_request, pod, self.k8s_api_request_timeout),
+            f'Could not create pod {self.pod_name}',
+            # Each req should be given k8s_api_request_timeout seconds.
+            # FIXME: We should instead add a timeout_times property to exponential_backoff instead
+            timeout=self.k8s_api_request_timeout * self.k8s_api_request_timeout_retries
```

Comment:
See above suggestion for making this timeout directly configurable instead of trying to calculate it from countable retries